pkg/controller/node/cidr_set.go

```diff
@@ -67,41 +82,70 @@ func (s *cidrSet) allocateNext() (*net.IPNet, error) {
 
 	s.used.SetBit(&s.used, nextUnused, 1)
 
-	j := uint32(nextUnused) << uint32(32-s.subNetMaskSize)
-	ipInt := (binary.BigEndian.Uint32(s.clusterIP)) | j
-	ip := make([]byte, 4)
-	binary.BigEndian.PutUint32(ip, ipInt)
+	var ip []byte
+	var mask int
 
+	switch /*v4 or v6*/ {
+	case s.clusterIP.To4() != nil:
+		{
+			j := uint32(nextUnused) << uint32(32-s.subNetMaskSize)
+			ipInt := (binary.BigEndian.Uint32(s.clusterIP)) | j
+			ip = make([]byte, 4)
+			binary.BigEndian.PutUint32(ip, ipInt)
+			mask = 32
+
+		}
+	case s.clusterIP.To16() != nil:
+		{
+			j := uint64(nextUnused) << uint64(64-s.subNetMaskSize)
+			ipInt := (binary.BigEndian.Uint64(s.clusterIP)) | j
+			ip = make([]byte, 16)
+			binary.BigEndian.PutUint64(ip, ipInt)
+			mask = 128
+		}
+	}
 	return &net.IPNet{
 		IP:   ip,
-		Mask: net.CIDRMask(s.subNetMaskSize, 32),
+		Mask: net.CIDRMask(s.subNetMaskSize, mask),
 	}, nil
 }
 
 func (s *cidrSet) getBeginingAndEndIndices(cidr *net.IPNet) (begin, end int, err error) {
 	begin, end = 0, s.maxCIDRs-1
 	cidrMask := cidr.Mask
 	maskSize, _ := cidrMask.Size()
+	var ipSize int
+	if cidr.IP.To4() != nil {
+		ipSize = net.IPv4len
+	} else if cidr.IP.To16() != nil {
+		ipSize = net.IPv6len
+	} else {
```

Comment:
I don't think the cidr can be invalid, it was more defensive. An invalid CIDR would error when it was parsed. Or maybe just check for cidr == nil.